subprojects/core/src/main/java/org/gradle/internal/model/StateTransitionControllerFactory.java

```diff
@@ -17,19 +17,36 @@
 package org.gradle.internal.model;
 
 import org.gradle.internal.DisplayName;
+import org.gradle.internal.operations.BuildOperationRunner;
+import org.gradle.internal.operations.BuildOperationsParameters;
 import org.gradle.internal.service.scopes.Scope;
 import org.gradle.internal.service.scopes.ServiceScope;
+import org.gradle.internal.work.Synchronizer;
+import org.gradle.internal.work.WaitBuildOperationFiringSynchronizer;
 import org.gradle.internal.work.WorkerLeaseService;
 
 @ServiceScope(Scope.BuildSession.class)
 public class StateTransitionControllerFactory {
+
     private final WorkerLeaseService workerLeaseService;
+    private final BuildOperationsParameters buildOperationsParameters;
+    private final BuildOperationRunner buildOperationRunner;
 
-    public StateTransitionControllerFactory(WorkerLeaseService workerLeaseService) {
+    public StateTransitionControllerFactory(
+        WorkerLeaseService workerLeaseService,
+        BuildOperationsParameters buildOperationsParameters,
+        BuildOperationRunner buildOperationRunner
+    ) {
         this.workerLeaseService = workerLeaseService;
+        this.buildOperationsParameters = buildOperationsParameters;
+        this.buildOperationRunner = buildOperationRunner;
     }
 
     public <T extends StateTransitionController.State> StateTransitionController<T> newController(DisplayName displayName, T initialState) {
-        return new StateTransitionController<>(displayName, initialState, workerLeaseService.newResource());
+        Synchronizer synchronizer = workerLeaseService.newResource();
+        if (buildOperationsParameters.isVerbose()) {
```

Comment:
❓ Could this decision be made once at factory instantiation time?

Reply:
I don't feel strongly here. I'd also prefer to save the flag in the field. But previously I received feedback that we want to delay decisions like that to the use sites. E.g. for usages of `BuildModelParameters` and similar classes.

Would you rather I change this?

Reply:
> But previously I received feedback that we want to delay decisions like that to the use sites

It would be interesting to recover the rationale there.

I can see how allowing dynamic reconfiguration of parts of Gradle might be interesting but, in general, it seems parameters / feature flags are frozen at initialization.

> Would you rather I change this?

As compiler buff, I'd have a slight preference for introducing a `VerboseStateTransitionControllerFactory` and switching at factory instantiation time which factory to use.

Reply:
I recall this comment about making build features lazily computed, even though they are essentially flags fixed during initialization. It seems to have been provided as a suggestion without additional context:
https://github.com/gradle/gradle/pull/26325#discussion_r1344060629

Reply:
I don't see how that comment about lazily instantiating some type related to my comment about making the decision about which strategy to use only once in this context.